ES|QL CCS uses skip_unavailable setting for handling disconnected remote clusters

quux00 opened the pull request:
As part of ES|QL planning of a cross-cluster search, a field-caps call is done to each cluster and,
if an ENRICH command is present, the enrich policy-resolve API is called on each remote. If a
remote cluster cannot be connected to in these calls, the outcome depends on the
skip_unavailable setting.

For skip_unavailable=false clusters, the error is fatal and the error will immediately be propagated
back to the client with a top level error message with a 500 HTTP status response code.

For skip_unavailable=true clusters, the error is not fatal. The error will be trapped, recorded in the
EsqlExecutionInfo object for the query, marking the cluster as SKIPPED. If the user requested
CCS metadata to be included, the cluster status and connection failure will be present in the
_clusters/details section of the response.

If no clusters can be contacted, if they are all marked as skip_unavailable=true, no error will be
returned. Instead a 200 HTTP status will be returned with no column and no values. If the
include_ccs_metadata: true setting was included on the query, the errors will listed in the
_clusters metadata section. (Note: this is also how the _search endpoint works for CCS.)

Partially addresses https://github.com/elastic/elasticsearch/issues/114531

quux00 committed:
Intmd commit in order to work on backports

quux00 committed:
Added LogicalPlanActionListener and improved edge case handling

quux00 committed:
Placeholder to start on enrich policy pathways

quux00 committed:
Intmd commit - very basic testing shows that this handles skip_un correctly; needs cleanup and more testing

quux00 committed:
Added Map<String, Exception> unavailableClusters to EnrichResolution

Created ConnectionErrorLookupResponse as subclass of LookupResponse

quux00 committed:
Completed first pass handling of remote unavailable errors during enrich policy API handling.

One key change: if after the enrich policy resolution we know we could not contact one or more clusters,
we remove them from the index expression that is sent to field-caps so we don't trigger another
round of network errors and ExecutionInfo updates.

quux00 committed:
Remove ConnectionErrorLookupResponse and add errorConnection to LookupResponse

quux00 committed:
Non-working version of CrossClusterQueryUnavailableIT added to get feedback

quux00 committed:
Working version of CrossClusterQueryUnavailableRemotesIT

quux00 committed:
Renamed to CrossClusterQueryUnavailableIT to CrossClusterQueryUnavailableRemotesIT

quux00 committed:
Enhancements to CrossClusterQueryUnavailableRemotesIT - not yet doing enrich API calls though

quux00 committed:
Fixed bug in EsqlSession found by existing tests and patched RemoteClusterSecurityEsqlIT to match the new error output

quux00 committed:
Added CrossClusterEnrichUnavailableClustersIT

Fixed some bugs in EnrichPolicyResolver and/or adjusted logic around handling missing remotes
in enrich based on testing.

quux00 committed:
Removed two tests from CrossClusterEnrichUnavailableClustersIT

quux00 committed:
Updated RemoteClusterSecurityEsqlIT to match new skip_unavailable=true error swallowing functionality

quux00 committed:
Minor cleanup before PR review

quux00 committed:
Moved updateExecutionInfoWithClustersWithNoMatchingIndices out of the indexResolution.isValid block

Other minor cleanup

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/115266.yaml

quux00 committed:
Removed change to jdk-deprecated.txt

quux00 committed:
Reversed the decision to pull updateExecutionInfoWithClustersWithNoMatchingIndices out of valid IndexResolution section

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,106 @@ public void execute(
+                return false;
+            }
+
+            if ((e instanceof IllegalStateException ill && ill.getMessage().equals("No clusters to search"))
```
Maybe it's better to have a special exception type for this? Matching by exception text seems a bit brittle.

quux00 replied:
Fixed.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -288,10 +403,34 @@ private void preAnalyzeIndices(
+                            clusterAlias,
+                            indexExpr,
+                            executionInfo.isSkipUnavailable(clusterAlias),
+                            EsqlExecutionInfo.Cluster.Status.SKIPPED,
```
Here `executionInfo.isSkipUnavailable(clusterAlias)` should actually always be true, right?

quux00 replied:
Yes.

smalyshev commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -288,10 +403,34 @@ private void preAnalyzeIndices(
+            // if the preceding call to the enrich policy API found unavailable clusters, recreate the index expression to search
+            // based only on available clusters (which could now be an empty list)
+            String indexExpressionToResolve = createIndexExpressionFromAvailableClusters(executionInfo);
+            if (indexExpressionToResolve.equals("")) {
```
why not `isEmpty`?

quux00 replied:
Fixed in next push.

smalyshev commented on x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityEsqlIT.java:
```diff
@@ -491,6 +495,10 @@ public void testCrossClusterQueryWithRemoteDLSAndFLS() throws Exception {
     }
 
+    /**
+     * Note: invalid_remote is "invalid" because it has a bogus API key, but the cluster does actually exist
```
Does `invalid_remote` also always have `skip_unavailable=true`? Because as I understand if it's not, it should error out?

quux00 replied:
Good catch. I think my understanding of this test is wrong. I'll work on it next and update the comments and make sure this is behaving as expected.

quux00 replied:
I've updated the javadoc comment. The test is actually missing and we get different errors from the security layer depending on its skip_un=true/false flag, which explains the different behavior in the test seen below (for the remote-only test; the local+remote test has a different issue).

smalyshev commented on x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityEsqlIT.java:
```diff
@@ -512,22 +520,49 @@ public void testCrossClusterQueryAgainstInvalidRemote() throws Exception {
+        {
+            var q = "FROM invalid_remote:employees,employees | SORT emp_id DESC | LIMIT 10";
+            Response response = performRequestWithRemoteSearchUser(esqlRequest(q));
+            assertLocalOnlyResults(response);
```
For example, here - if `skipUnavailable` is false, shouldn't it be an error here?

quux00 replied:
I analyzed this case and the problem is that field-caps is returning nothing from this cluster, rather than an Exception, so it is indistinguishable from the scenario where no matching indices could be found. The "no matching indices" found is going to be handled in a follow-on PR, so I will investigate this tests behavior under RCS 2.0 in that test.

quux00 committed:
Adjusted EsqlSessionTests to new logic of zeroing out shard counts for SKIPPED clusters

smalyshev commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterEnrichUnavailableClustersIT.java:
```diff
@@ -0,0 +1,736 @@
+        return Settings.builder().put(super.nodeSettings()).put(XPackSettings.SECURITY_ENABLED.getKey(), false).build();
+    }
+
+    static final EnrichPolicy hostPolicy = new EnrichPolicy("match", null, List.of("hosts"), "ip", List.of("ip", "os"));
```
These policies seem to be the same as in other test class - maybe we should take them out into an utility class?

quux00 replied:
Thanks. I changed this test to use several methods and static members from the CrossClustersEnrichIT.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

quux00 committed:
PR feedback

DRY'd up CrossClusterEnrichUnavailableClustersIT to reuse set up from CrossClustersEnrichIT
Created NoClustersToSearchException sentinel exception
Clarified how RemoteClusterSecurityEsqlIT works and flagged it for more work in a follow-on PR.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

astefan reviewed (commented):
That's great @quux00.
I've left some comments.
I was also about to add something related to testing scenarios involving inexistent indices (excluding the remote clusters not being available case), but I think you are going to cover those in the second PR right?

For example, the case of `from remote:existent, inexistent*` where there should be no error and the case of `from remote:existent, inexistent` where there should be an error. This is not strictly related to CCS and `skip_unvailable` but it would be good to see that the above behavior involving missing local indices (with and without wildcard) is also tested.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java:
```diff
@@ -261,22 +281,34 @@ private void lookupPolicies(
             if (remotePolicies.isEmpty() == false) {
                 for (String cluster : remoteClusters) {
```
No, the enrich policy resolution is done before the field-caps. I'm not sure why or if that's necessity or just historical accident. IMO, the field-caps call seems the logical one to do first, but that's not how it is, so we need to track unavailable clusters here and then pass that list to the callback listener which is the one that does the field-caps lookup: https://github.com/elastic/elasticsearch/pull/115266/files#diff-40060e2ec9003953a228c4a03bdc80a301949b0d4e3dccc1978798e33a992a73R345

smalyshev replied:
Oh ok, I was for some reason assuming field caps go first... then of course yes, enrich should be telling field caps which clusters are skipped, not the other way around.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,106 @@ public void execute(
+                    // it just throws the first exception in its list, so this odd special handling is here is to avoid
+                    // having one specific remote alias name in all failure lists in the metadata response
+                    exceptionForResponse = new RemoteTransportException(
+                        "connect_transport_exception - unable to connect to remote cluster",
```
If you have two clusters, both skip_un=true, and field-caps cannot connect to either of them, rather than returning both errors, field-caps returns one (arbitrarily). If you just put that error message in the ExecutionInfo.Cluster, you get this incorrect and confusing output:

```
 "_clusters": {
    "total": 2,
    "successful": 0,
    "running": 0,
    "skipped": 2,
    "partial": 0,
    "failed": 0,
    "details": {
      "remote2": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "connect_transport_exception",
              "reason": "[could not connect to remote1]"
            }
          }
        ]
      },
      "remote1": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "connect_transport_exception",
              "reason": "[could not connect to remote1]"
            }
          }
        ]
      }
    }
```

This special handling changes it to be a generic "could not connect to remote cluster" message, like so:

```
 "_clusters": {
    "total": 2,
    "successful": 0,
    "running": 0,
    "skipped": 2,
    "partial": 0,
    "failed": 0,
    "details": {
      "remote2": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      },
      "remote1": {
        "status": "skipped",
        "indices": "foo",
        "took": 96,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      }
    }
```

quux00 replied:
> This message is a bit confusing to me - is it about connecting to some specific cluster
The error message occurs within the cluster/details of a specific cluster, so the intention is that it is obvious which cluster we could not connect to.

I could add additional handling to swap in the name of each cluster into the error message if you think this model is not clear.

smalyshev replied:
No if it's inside specific cluster's data then it's fine I think.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java:
```diff
@@ -149,13 +166,16 @@ private Tuple<ResolvedEnrichPolicy, String> mergeLookupResults(
-        assert targetClusters.isEmpty() == false;
         String policyName = unresolved.name;
+        if (targetClusters.isEmpty()) {
+            return Tuple.tuple(null, "enrich policy [" + policyName + "] cannot be resolved since remote clusters are unavailable");
```
This code was added in response to one of the IT tests I added failing:
CrossClusterEnrichUnavailableClustersIT..testEnrichWithHostsPolicyAndDisconnectedRemotesWithSkipUnavailableTrue, and in particular when the 2nd cluster is taken offline: https://github.com/elastic/elasticsearch/pull/115266/files#diff-67efeda55a8a497d595e7ce7737ee3ae3acc3ec966db031190bd6954a610f88aR240

In that case invariants in the code below are not met and assertion failures are thrown, so I added this to just leave the code immediately when there are no targetClusters, which can happen when the Enrich Mode type is "REMOTE" and the local cluster was not included in the query.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,106 @@ public void execute(
+                            executionInfo.overallTook()
+                        ).setTotalShards(0).setSuccessfulShards(0).setSkippedShards(0).setFailedShards(0);
+                        if (RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY.equals(clusterAlias)) {
+                            // never mark local cluster as skipped
```
Example where this is needed:

```
  "query": "from noindex*,remote1:employees,remote2:noindex*|\n STATS count(*)",
```

where remote1 is down (unavailable) and both remote clusters are `skip_un=true`.
What happens is that since there are no indexes to search on any of the requested clusters and since both remotes were skip_un=true and since the local cluster indices were specified with a wildcard (lenient handling), we don't throw a fatal exception.

But in discussion with Andrei, the local cluster will never be marked as skipped. We may also change this to be a fatal scenario in the follow on ticket where we will deal with "missing indices" failures, so for now the rule is just to not mark local as skipped.

For the above case, without the special handling in this code, you get incorrect information in the "(local)" cluster/details entry: 
```
  "_clusters": {
    "total": 3,
    "successful": 0,
    "running": 0,
    "skipped": 3,
    "partial": 0,
    "failed": 0,
    "details": {
      "(local)": {
        "status": "skipped",
        "indices": "noindex*",
        "took": 18,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      },
      "remote2": {
        "status": "skipped",
        "indices": "noindex*",
        "took": 18,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      },
      "remote1": {
        "status": "skipped",
        "indices": "employees",
        "took": 18,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      }
    }
  }
```

quux00 replied:
I didn't have a test for this case, so I've added one in CrossClusterQueryUnavailableRemotesIT on the next push.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,106 @@ public void execute(
+        /**
+         * Whether to return an empty result (HTTP status 200) for a CCS rather than a top level 4xx/5xx error.
+         *
+         * For cases where field-caps had no indices to search and the remotes were unavailable, we
```
First comment - I was planning to handle the "missing indices" questions in "ticket 2" of the skip_unavailable work (this is ticket 1 dealing with just disconnected clusters).

That said, let's clarify here what behavior you are proposing:

1) If, at the end of cluster resolution (and checking for connectivity) and index resolution, there are no indices to query (either because the clusters are unavailable and/or the available clusters have no matching indices), then we should throw a VerificationException, rather than an empty "successful" result.

2) The above is true even if it is a remote-only query and wildcards were used for either the cluster names or the index names or both.

If I have that right - do you think there is any scenario where **at planning time** (before starting ComputeService execution) we should return an empty "successful" result?

astefan replied:
> 1. If, at the end of cluster resolution (and checking for connectivity) and index resolution, there are no indices to query (either because the clusters are unavailable and/or the available clusters have no matching indices), then we should throw a VerificationException, rather than an empty "successful" result.

No :-). My comment was specific to `from inexistent, remote:inexistent` and `from inexistent, remote:inexistent*`. This is a scenario that I missed to cover when we discussed about behavior of CCS queries. This involves the **local cluster** as well and I _think_ in this case it doesn't matter how CCS behaves. The fact that the **local cluster doesn't have an explicitly specified index** should fail no matter what. In the two tables from https://github.com/elastic/elasticsearch/issues/112886 (skip_un true/false) a better comparison query than the one in my previous review comment is 
> from remote:existent, inexistent

where the behavior is 

> failure coming from ES at query time complaining about "inexistent" from the "local" cluster

But, in our case the failure should be coming from the Verifier because the `field_caps` call returned no indices.

Another two use cases are `from inexistent*, remote:inexistent` and `from inexistent*, remote:inexistent*` and the principle for both these two cases and the previous two should be _if the local cluster is involved and the overall - remote and local - resolution of indices ends up with no indices this should be an error because the user specified a missing indices pattern for the local cluster; even if the user cannot control how the remote clusters behave, the user is in control of the index pattern for the local cluster_.

**skip_unavailable = true/false**

|index pattern                   | behavior                                                                                                                     |
|----------------------------|------------------------------------------------------------------------------------------|
| from inexistent*, remote:inexistent | failure     coming from post-`_field_caps` check, at the `Verifier` level |
| from inexistent*, remote:inexistent*  |  failure     coming from post-`_field_caps` check, at the `Verifier` level |

ESQL's current behavior where **no CCS** is used is to reject queries that have no indices after the index pattern is resolved:
`from inexistent*`
`from inexistent1*,inexistent2*`
`from inexistent`
`from inexistent1,inexistent2`
`from inexistent,inexistent1*`

I have updated the tables from https://github.com/elastic/elasticsearch/issues/112886

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,106 @@ public void execute(
+         * on any of the requested clusters.
+         */
+        private boolean returnSuccessWithEmptyResult(Exception e) {
+            if (executionInfo.isCrossClusterSearch() == false) {
```
Sorry, I don't think I follow what you are saying here. It sounds like there is a scenario where you think we should return an empty successful result at planning. Could you outline what that is again? (Same question I just asked in the previous comment, so you can just answer there.)

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterEnrichUnavailableClustersIT.java:
```diff
@@ -0,0 +1,736 @@
+        return false;
+    }
+
+    protected Collection<String> allClusters() {
```
Fixed.

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterEnrichUnavailableClustersIT.java:
```diff
@@ -0,0 +1,736 @@
+                String query = "FROM *:events | eval ip= TO_STR(host) | " + enrichHosts(mode) + " | stats c = COUNT(*) by os | SORT os";
+                try (EsqlQueryResponse resp = runQuery(query, requestIncludeMeta)) {
+                    List<List<Object>> rows = getValuesList(resp);
+                    assertThat(rows.size(), equalTo(0));
```
Oh, interesting. Thanks for flagging that. I'll work on adding that next. I've seen that response in my manual testing and didn't know what it meant.

quux00 replied:
OK, I added it and did some manual testing. So here's a remote-only test against `remote1` which is offline and skip_unavailable true. Is this the output that looks best to you?

```
{
  "is_running": false,
  "took": 7,
  "all_columns": [
    {
      "name": "<no-fields>",
      "type": "null"
    }
  ],
  "columns": [],
  "values": [],
  "_clusters": {
    "total": 1,
    "successful": 0,
    "running": 0,
    "skipped": 1,
    "partial": 0,
    "failed": 0,
    "details": {
      "remote1": {
        "status": "skipped",
        "indices": "bl*",
        "took": 7,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        },
        "failures": [
          {
            "shard": -1,
            "index": null,
            "reason": {
              "type": "remote_transport_exception",
              "reason": "[connect_transport_exception - unable to connect to remote cluster]"
            }
          }
        ]
      }
    }
  }
}
```

quux00 replied:
Added that behavior and modified two tests in this IT test in [this commit](https://github.com/elastic/elasticsearch/pull/115266/commits/dc2a75b03e4878fcad629baf6242ced8c82d438b).

quux00 committed:
PR feedback. Added additional test to CrossClusterQueryUnavailableRemotesIT. Some minor code cleanup

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

quux00 committed:
PR feedback: Empty Results (skip_un=true scenarios) now return Analyzer.NO_FIELDS as the schema; tests updated

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterQueryUnavailableRemotesIT.java:
```diff
@@ -0,0 +1,460 @@
+                assertClusterMetadataInResponse(resp, responseExpectMeta);
+            }
+
+            // close remote-cluster-12that it is also unavailable
```
Thanks. Fixed.

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterQueryUnavailableRemotesIT.java:
```diff
@@ -0,0 +1,460 @@
+            // query only the REMOTE_CLUSTER_1
+            try (EsqlQueryResponse resp = runQuery("FROM " + REMOTE_CLUSTER_1 + ":logs-* | STATS sum (v)", requestIncludeMeta)) {
+                List<List<Object>> values = getValuesList(resp);
+                assertThat(values, hasSize(0));
```
Fixed in next push.

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterQueryUnavailableRemotesIT.java:
```diff
@@ -0,0 +1,460 @@
+            // close remote cluster 2 so that it is also unavailable
+            cluster(REMOTE_CLUSTER_2).close();
+
+            // query only the REMOTE_CLUSTER_1
```
Thanks. Fixed.

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterQueryUnavailableRemotesIT.java:
```diff
@@ -0,0 +1,460 @@
+        clusterInfo.put("remote.index", remoteIndex);
+
+        if (numClusters == 3) {
+            String remoteIndex2 = "logs-2";
```
Good idea. The original only had one cluster and I didn't update the local var name here. Will fix in next push.

quux00 replied:
Fixed.

quux00 committed:
Added checks for no-fields columns to CrossClusterQueryUnavailableRemotesIT when empty result in returned

quux00 committed:
Fix failing test and spotless check

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

quux00 committed:
Minor cleanup

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

astefan reviewed (approved):
LGTM
Thanks @quux00

quux00 commented on x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/CrossClusterQueryUnavailableRemotesIT.java:
```diff
@@ -135,7 +136,60 @@ public void testCCSAgainstDisconnectedRemoteWithSkipUnavailableTrue() throws Exc
+            // 1) the local cluster indexExpression and REMOTE_CLUSTER_2 indexExpression match no indices
+            // 2) the REMOTE_CLUSTER_1 is unavailable
+            // 3) both remotes are marked as skip_un=true
+            String query = "FROM nomatch*," + REMOTE_CLUSTER_1 + ":logs-*," + REMOTE_CLUSTER_2 + ":nomatch* | STATS sum (v)";
```
Correct!

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql-ccs/skip_unavailable-disconnected-t1

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlExecutionInfo.java:
```diff
@@ -349,6 +352,11 @@ public Cluster(
             this.successfulShards = successfulShards;
             this.skippedShards = skippedShards;
             this.failedShards = failedShards;
+            if (failures == null) {
+                this.failures = List.of();
+            } else {
+                this.failures = failures;
+            }
```
`this.failures = failures == null ? emptyList() : failures`

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlExecutionInfo.java:
```diff
@@ -362,6 +370,11 @@ public Cluster(StreamInput in) throws IOException {
             this.took = in.readOptionalTimeValue();
             this.skipUnavailable = in.readBoolean();
+            if (in.getTransportVersion().onOrAfter(TransportVersions.ESQL_CCS_EXEC_INFO_WITH_FAILURES)) {
+                this.failures = Collections.unmodifiableList(in.readCollectionAsList(ShardSearchFailure::readShardSearchFailure));
```
Any reason why the collection is unmodifiable - within the ESQL code we're not using the defensive style since the collections are never modified, rather copied.

quux00 replied:
The EsqlExecution.Cluster class, like the SearchResponse.Cluster class it is based on, is immutable. You can only change state by coping an existing Cluster and swapping it in. I can change line 374 to use a mutable list, but it won't make any difference since the coding practice for this class is to never modify in place. In fact, the line below where I do `this.failures = List.of();` should likely be changed to an immutable list as well to enforce this model.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java:
```diff
@@ -261,22 +281,34 @@ private void lookupPolicies(
-                            )
-                        )
-                    );
+                                new ActionListenerResponseHandler<>(lookupListener.delegateResponse((l, e) -> {
+                                    if (ExceptionsHelper.isRemoteUnavailableException(e)
+                                        && remoteClusterService.isSkipUnavailable(cluster)) {
+                                        l.onResponse(new LookupResponse(e));
+                                    } else {
+                                        l.onFailure(e);
+                                    }
+                                }), LookupResponse::new, threadPool.executor(ThreadPool.Names.SEARCH))
+                            );
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            if (ExceptionsHelper.isRemoteUnavailableException(e) && remoteClusterService.isSkipUnavailable(cluster)) {
+                                lookupListener.onResponse(new LookupResponse(e));
+                            } else {
+                                lookupListener.onFailure(e);
+                            }
+                        }
+                    });
```
Since the code is similar between onFailure and onResponse, pull it out in a separate method.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,105 @@ public void execute(
+     * Any Exception sent to onFailure stops processing, but not all are fatal (return a 4xx or 5xx), so
+     * the onFailure handler determines whether to return an empty successful result or a 4xx/5xx error.
+     */
+    class LogicalPlanActionListener implements ActionListener<LogicalPlan> {
```
As far as I can tell, the reason for this listener is to pick up remote CCS error and properly report them. These occur during the preAnalysis phase or the execution.
Hence why I suggest to:
- use a different name that better conveys the purpose (e.g. CcsAwareActionListener or CcsPartialFailuresActionListener)
- update the javadoc description accordingly

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -143,12 +150,105 @@ public void execute(
+
+        @Override
+        public void onFailure(Exception e) {
+            if (returnSuccessWithEmptyResult(e)) {
```
To improve readability, please extract this code into a CcsUtils class and call it here as a one liner:
```
CcsUtils.handleException(Exception, ExecutionInfo, ActionListener)
```

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -302,6 +440,30 @@ private void preAnalyzeIndices(
     }
 
+    // visible for testing
+    static String createIndexExpressionFromAvailableClusters(EsqlExecutionInfo executionInfo) {
```
Move this into a separate class CcsUtils

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -446,14 +608,28 @@ public PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan) {
-            executionInfo.swapCluster(
-                clusterAlias,
-                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SKIPPED).build()
+    static void updateExecutionInfoWithUnavailableClusters(EsqlExecutionInfo execInfo, Map<String, FieldCapabilitiesFailure> unavailable) {
```
Move this into a separate CcsUtils

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -158,18 +159,18 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
     // visible for testing
-    static Set<String> determineUnavailableRemoteClusters(List<FieldCapabilitiesFailure> failures) {
-        Set<String> unavailableRemotes = new HashSet<>();
+    static Map<String, FieldCapabilitiesFailure> determineUnavailableRemoteClusters(List<FieldCapabilitiesFailure> failures) {
```
Same as above- look into moving this into a CcsUtils class

costin reviewed (commented):
I'm leaving feedback here after this PR was merged for future enhancement (as I understand there's going to be more work).

It mainly resolved around the Enrich and CCS intricacies being spread through-out EsqlSession which makes evolving this problematic  - as this class is at the code of doing cluster calls.

The Enrich and CCS parts need encapsulation (such as moving to a separate class like CcsUtils, defining a dedicated listener, etc...) so that EsqlSession focused on its core purpose : working on the LogicalPlan.
Please make this a priority moving forward - thanks!
